### ISSUE TITLE
Each checkpoint at specified epoch overwrites the last

### ISSUE THREAD

**Issue by Graylington** [rhasspy/piper#317]
When training, I set "--checkpoint-epochs 100" so that I'll have a model saved every 100 epochs. While it does save the model each time, it overwrites the existing one from the previous 100 epoch checkpoint.

How can I make it keep every checkpoint along the way?

**Comment by jereskuta** [comment]
Having this issue as well.

**Comment by aaronnewsome** [comment]
Yes, me too. The command line options that I thought would keep checkpoints periodically don't seem to do anything. They always get removed when a new checkpoint is created. I set up a "while true" bash loop in a separate terminal to copy the checkpoints periodically. Seems to work well since everything is on NVME disk and the copy happens in a snap.

**Comment by Graylington** [comment]
My own solution for this in Windows, was to run a python script that continuously monitors the WSL Linux directory you chose for your checkpoints, and when it finds a new checkpoint without an existing backup it creates one alongside the original checkpoint file.

**_Run.bat**
`
REM Run.bat
@echo off
C:\Users\Graylington\AppData\Local\Programs\Python\Python310\python.exe WSLCheckpointBackupMonitor.py
pause
`

**WSLCheckpointBackupMonitor.py**
`
import os
import shutil
import time
from datetime import datetime

# Function to print the list of .ckpt files and create backups
def check_and_backup_files(directory):
    # Get the current time and format it as a string
    current_time = datetime.now().strftime("%H:%M:%S")
    
    # Initialize a flag to track whether a new backup is created
    new_backup_created = False
    
    # List files in the specified directory, prepended with the timestamp
    # print(f"{current_time} - Checking for .ckpt files in '{directory}'")
    try:
        for file in os.listdir(directory):
            # Check if the file has the .ckpt extension
            if file.endswith(".ckpt"):
                full_file_path = os.path.join(directory, file)
                backup_file_path = full_file_path + ".bak"
                
                # Check if the backup file already exists
                if not os.path.exists(backup_file_path):
                    # If it doesn't, create a backup
                    shutil.copy(full_file_path, backup_file_path)
                    print(f"{current_time} - Backup created for: {file}")
                    new_backup_created = True
                # else:
                    # print(f"{current_time} - Backup already exists for: {file}")
                    
        if not new_backup_created:
            # If no new backup was created, report that no new .ckpt files were found
            print(f"{current_time} - No new checkpoints requiring backup found.")
            
    except FileNotFoundError:
        print(f"{current_time} - Directory not found. Please check the path.")
    print("------------------------------------------------")

# Corrected UNC path to access the WSL directory
wsl_directory = r"\\wsl.localhost\Ubuntu\home\graylington\piper\my-training-charlotte\lightning_logs\version_0\checkpoints"

# Main loop to keep checking the directory every 5 seconds
while True:
    check_and_backup_files(wsl_directory)
    # Wait for 10 seconds before rechecking
    time.sleep(10)
`
Remember to update the wsl_directory to match your own path.

**Comment by Graylington** [comment]
My own solution for this in Windows was to run a python script that continuously monitors the WSL Linux directory you chose for your checkpoints, and when it finds a new checkpoint without an existing backup it creates one alongside the original checkpoint file.

**StartMonitoring.bat**
```
REM Run.bat
@echo off
C:\Users\Graylington\AppData\Local\Programs\Python\Python310\python.exe WSLCheckpointBackupMonitor.py
pause
```

**WSLCheckpointBackupMonitor.py**
*Remember to update the _wsl_directory_ to match your own path.
```
import os
import shutil
import time
from datetime import datetime

# Function to print the list of .ckpt files and create backups
def check_and_backup_files(directory):
    # Get the current time and format it as a string
    current_time = datetime.now().strftime("%H:%M:%S")
    
    # Initialize a flag to track whether a new backup is created
    new_backup_created = False
    
    # List files in the specified directory, prepended with the timestamp
    # print(f"{current_time} - Checking for .ckpt files in '{directory}'")
    try:
        for file in os.listdir(directory):
            # Check if the file has the .ckpt extension
            if file.endswith(".ckpt"):
                full_file_path = os.path.join(directory, file)
                backup_file_path = full_file_path + ".bak"
                
                # Check if the backup file already exists
                if not os.path.exists(backup_file_path):
                    # If it doesn't, create a backup
                    shutil.copy(full_file_path, backup_file_path)
                    print(f"{current_time} - Backup created for: {file}")
                    new_backup_created = True
                # else:
                    # print(f"{current_time} - Backup already exists for: {file}")
                    
        if not new_backup_created:
            # If no new backup was created, report that no new .ckpt files were found
            print(f"{current_time} - No new checkpoints requiring backup found.")
            
    except FileNotFoundError:
        print(f"{current_time} - Directory not found. Please check the path.")
    print("------------------------------------------------")

# Corrected UNC path to access the WSL directory
wsl_directory = r"\\wsl.localhost\Ubuntu\home\graylington\piper\my-training-charlotte\lightning_logs\version_0\checkpoints"

# Main loop to keep checking the directory every 5 seconds
while True:
    check_and_backup_files(wsl_directory)
    # Wait for 10 seconds before rechecking
    time.sleep(10)
```
